File: modules/io/tests/test_io.cpp
/*
 * Copyright 2022 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
#include <lagrange/IndexedAttribute.h>
#include <lagrange/Logger.h>
#include <lagrange/SurfaceMeshTypes.h>
#include <lagrange/attribute_names.h>
#include <lagrange/fs/filesystem.h>
#include <lagrange/io/load_mesh.h>
#include <lagrange/io/save_mesh.h>
#include <lagrange/testing/common.h>
#include <lagrange/utils/safe_cast.h>

// clang-format off
#include <lagrange/utils/warnoff.h>
#include <spdlog/fmt/fmt.h>
#include <catch2/catch_test_macros.hpp>
#include <catch2/benchmark/catch_benchmark.hpp>
#include <lagrange/utils/warnon.h>
// clang-format on

#include <fstream>
#include <random>

namespace fs = lagrange::fs;

namespace {

template <typename S, typename I>
void test_load_save()
{
    using namespace lagrange;
    using MeshType = SurfaceMesh<S, I>;
    using Index = typename MeshType::Index;

    std::mt19937 gen;
    std::uniform_real_distribution<S> dist(0, 1);

    for (int i = 0; i < 8; ++i) {
        fs::path filename = fmt::format("semi{}.obj", i + 1);
        auto mesh = lagrange::testing::load_surface_mesh<S, I>("open/core/tilings" / filename);
        logger().info(
            "Loaded tiling with {} vertices and {} facets",
            mesh.get_num_vertices(),
            mesh.get_num_facets());
        REQUIRE(mesh.is_hybrid());
        for (Index v = 0; v < mesh.get_num_vertices(); ++v) {
            mesh.ref_position(v)[2] = dist(gen);
        }
        fs::path output_path = lagrange::testing::get_test_output_path("test_io" / filename);
        io::save_mesh(output_path, mesh);
    }

    std::array<std::pair<std::string, int>, 3> test_cases = {
        {{"hexagon", 6}, {"square", 4}, {"triangle", 3}}};
    for (auto kv : test_cases) {
        fs::path filename = fmt::format("{}.obj", kv.first);
        auto mesh = lagrange::testing::load_surface_mesh<S, I>("open/core/tilings" / filename);
        logger().info(
            "Loaded tiling with {} vertices and {} facets",
            mesh.get_num_vertices(),
            mesh.get_num_facets());
        REQUIRE(mesh.is_regular());
        REQUIRE(mesh.get_vertex_per_facet() == safe_cast<Index>(kv.second));
        for (Index v = 0; v < mesh.get_num_vertices(); ++v) {
            mesh.ref_position(v)[2] = dist(gen);
        }
        fs::path output_path = lagrange::testing::get_test_output_path("test_io" / filename);
        io::save_mesh(output_path, mesh);
    }
}

template <typename S, typename I>
void test_benchmark_tiles()
{
    using namespace lagrange;
    using MeshType = SurfaceMesh<S, I>;
    using Index = typename MeshType::Index;

    BENCHMARK("Timings Tiles")
    {
        int n = 0;

        for (int i = 0; i < 8; ++i) {
            fs::path filename = fmt::format("semi{}.obj", i + 1);
            auto mesh = lagrange::testing::load_surface_mesh<S, I>("open/core/tilings" / filename);
            n += safe_cast<int>(mesh.get_num_vertices());
            REQUIRE(mesh.is_hybrid());
            fs::path output_path = lagrange::testing::get_test_output_path("test_io" / filename);
            io::save_mesh(output_path, mesh);
        }

        std::array<std::pair<std::string, int>, 3> test_cases = {
            {{"hexagon", 6}, {"square", 4}, {"triangle", 3}}};
        for (auto kv : test_cases) {
            fs::path filename = fmt::format("{}.obj", kv.first);
            auto mesh = lagrange::testing::load_surface_mesh<S, I>("open/core/tilings" / filename);
            n += safe_cast<int>(mesh.get_num_vertices());
            REQUIRE(mesh.is_regular());
            REQUIRE(mesh.get_vertex_per_facet() == safe_cast<Index>(kv.second));
            fs::path output_path = lagrange::testing::get_test_output_path("test_io" / filename);
            io::save_mesh(output_path, mesh);
        }

        return n;
    };
}

template <typename S, typename I>
void test_io_blub()
{
    using namespace lagrange;
    using MeshType = SurfaceMesh<S, I>;
    using Scalar = typename MeshType::Scalar;

    auto mesh = lagrange::testing::load_surface_mesh<S, I>("open/core/blub/blub.obj");
    fs::path output_path = lagrange::testing::get_test_output_path("test_io/blub.obj");
    io::save_mesh(output_path, mesh);

    logger().info("Mesh #v {}, #f {}", mesh.get_num_vertices(), mesh.get_num_facets());
    auto& uv_attr = mesh.template get_indexed_attribute<Scalar>(AttributeName::texcoord);
    logger().info("Mesh #uv {}", uv_attr.values().get_num_elements());
}

template <typename S, typename I>
void test_benchmark_large()
{
    using namespace lagrange;
    using MeshType = SurfaceMesh<S, I>;

    BENCHMARK("Timings Large")
    {
        // TODO: Add this guy to our unit test data
        fs::path path = "/Users/jedumas/cloud/adobe/shared/mesh_processing/Modeler - "
                        "Qadremesher/SoylentGreen_FullRes.obj";
        auto mesh = io::load_mesh<MeshType>(path.string());
        logger().info("Mesh #v {}, #f {}", mesh.get_num_vertices(), mesh.get_num_facets());
        return mesh.get_num_vertices();
        // Uncomment to time obj save as well
        // fs::path output_path = lagrange::testing::get_test_output_path("test_io/large_out.obj");
        // fs::ofstream output_stream(output_path);
        // io::save_mesh_obj(output_stream, mesh);
    };
}

template <typename S, typename I>
void test_obj_indexing()
{
    using namespace lagrange;
    using MeshType = SurfaceMesh<S, I>;
    using Index = typename MeshType::Index;

    auto mesh = lagrange::testing::load_surface_mesh<S, I>("open/core/index-test.obj");

    for (Index f = 0; f < mesh.get_num_facets(); ++f) {
        const Index first_corner = mesh.get_facet_corner_begin(f);
        const Index last_corner = mesh.get_facet_corner_end(f);

        bool all_zero = true;
        for (Index c = first_corner; c < last_corner; ++c) {
            Index vertexIndex = mesh.get_corner_vertex(c);
            all_zero = all_zero && (vertexIndex == 0);
        }

        // Incorrect mesh indexing during obj load will result in uninitialized facets.
        REQUIRE(!all_zero);
    }
}

} // namespace

TEST_CASE("Mesh IO: Load and Save", "[next]")
{
#define LA_X_test_load_save(_, S, I) test_load_save<S, I>();
    LA_SURFACE_MESH_X(test_load_save, 0)
}

TEST_CASE("Mesh IO: Benchmark Tiles", "[next][!benchmark]")
{
#define LA_X_test_benchmark_tiles(_, S, I) test_benchmark_tiles<S, I>();
    LA_SURFACE_MESH_X(test_benchmark_tiles, 0)
}

TEST_CASE("Mesh IO: Blub", "[next]")
{
#define LA_X_test_io_blub(_, S, I) test_io_blub<S, I>();
    LA_SURFACE_MESH_X(test_io_blub, 0)
}

TEST_CASE("Mesh IO: Benchmark Large", "[next][!benchmark]" LA_CORP_FLAG)
{
#define LA_X_test_benchmark_large(_, S, I) test_benchmark_large<S, I>();
    LA_SURFACE_MESH_X(test_benchmark_large, 0)
}

TEST_CASE("Mesh IO: Index Test", "[next]")
{
#define LA_X_test_obj_indexing(_, S, I) test_obj_indexing<S, I>();
    LA_SURFACE_MESH_X(test_obj_indexing, 0)
}

File: modules/io/python/tests/test_io.py
#
# Copyright 2023 Adobe. All rights reserved.
# This file is licensed to you under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License. You may obtain a copy
# of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under
# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
# OF ANY KIND, either express or implied. See the License for the specific language
# governing permissions and limitations under the License.
#
import lagrange

import datetime
import numpy as np
import tempfile
import pathlib
import pytest


@pytest.fixture
def triangle():
    mesh = lagrange.SurfaceMesh()
    mesh.add_vertices(np.identity(3))
    mesh.add_triangle(0, 1, 2)
    return mesh


@pytest.fixture
def triangle_with_uv_normal_color(triangle):
    mesh = triangle
    uv_values = np.array([[0, 0], [1, 0], [0, 1]])
    uv_indices = np.array([[0, 1, 2]])
    mesh.create_attribute(
        "uv",
        lagrange.AttributeElement.Indexed,
        lagrange.AttributeUsage.UV,
        uv_values,
        uv_indices,
    )
    lagrange.compute_normal(mesh)

    color_values = np.array([[255, 255, 255]])
    color_indices = np.array([[0, 0, 0]])
    mesh.create_attribute(
        "color",
        lagrange.AttributeElement.Indexed,
        lagrange.AttributeUsage.Color,
        color_values,
        color_indices,
    )
    return mesh


def assert_same_vertices_and_facets(mesh, mesh2):
    assert mesh.num_vertices == mesh2.num_vertices
    assert mesh.num_facets == mesh2.num_facets
    assert mesh.vertices == pytest.approx(mesh2.vertices)
    assert np.all(mesh.facets == mesh2.facets)


def match_attribute(mesh, mesh2, id1, id2):
    attr_name = "__unit_test__"
    # Convert both attributes to corner attribute and compare the per-corner value.
    id1 = lagrange.map_attribute(
        mesh, id1, attr_name, lagrange.AttributeElement.Corner
    )
    id2 = lagrange.map_attribute(
        mesh2, id2, attr_name, lagrange.AttributeElement.Corner
    )

    attr1 = mesh.attribute(id1)
    attr2 = mesh2.attribute(id2)

    assert attr1.data == pytest.approx(attr2.data)
    mesh.delete_attribute(attr_name)
    mesh2.delete_attribute(attr_name)


def assert_same_attribute(mesh, mesh2, usage, required):
    elements = [
        lagrange.AttributeElement.Indexed,
        lagrange.AttributeElement.Vertex,
        lagrange.AttributeElement.Facet,
        lagrange.AttributeElement.Corner,
        lagrange.AttributeElement.Edge,
    ]

    ids_1 = []
    for elem in elements:
        ids_1 += mesh.get_matching_attribute_ids(element=elem, usage=usage)
    if len(ids_1) == 0:
        return

    ids_2 = []
    for elem in elements:
        ids_2 += mesh2.get_matching_attribute_ids(element=elem, usage=usage)

    assert not required or len(ids_2) > 0
    for id1 in ids_1:
        for id2 in ids_2:
            match_attribute(mesh, mesh2, id1, id2)


class TestIO:
    def __save_and_load__(
        self, suffix, mesh, binary, exact_match, selected_attributes
    ):
        with tempfile.TemporaryDirectory() as tmp_dir:
            tmp_dir_path = pathlib.Path(tmp_dir)
            stamp = datetime.datetime.now().isoformat().replace(":", ".")
            filename = tmp_dir_path / f"{stamp}{suffix}"
            lagrange.io.save_mesh(
                filename,
                mesh,
                binary=binary,
                exact_match=exact_match,
                selected_attributes=selected_attributes,
            )
            mesh2 = lagrange.io.load_mesh(filename)

            assert_same_vertices_and_facets(mesh, mesh2)

            required = not exact_match
            # Check special attributes.
            assert_same_attribute(
                mesh, mesh2, lagrange.AttributeUsage.UV, required
            )
            assert_same_attribute(
                mesh, mesh2, lagrange.AttributeUsage.Normal, required
            )
            if filename.suffix != ".obj":
                assert_same_attribute(
                    mesh, mesh2, lagrange.AttributeUsage.Color, required
                )

            if selected_attributes is not None:
                for attr_id in selected_attributes:
                    if mesh.is_attribute_indexed(attr_id):
                        pass
                    else:
                        attr_name = mesh.get_attribute_name(attr_id)
                        attr = mesh.attribute(attr_id)
                        assert mesh2.has_attribute(attr_name)

    def save_and_load(self, mesh, attribute_ids=None):
        for exact_match in [True, False]:
            # For ascii formats.
            for ext in [".obj", ".ply", ".msh"]:
                self.__save_and_load__(
                    ext,
                    mesh,
                    binary=False,
                    exact_match=exact_match,
                    selected_attributes=attribute_ids,
                )

            # For binary formats
            for ext in [".ply", ".msh"]:
                self.__save_and_load__(
                    ext,
                    mesh,
                    binary=True,
                    exact_match=exact_match,
                    selected_attributes=attribute_ids,
                )

    def test_single_triangle(self, triangle):
        mesh = triangle
        self.save_and_load(mesh)

    def test_single_triangle_with_attributes(
        self, triangle_with_uv_normal_color
    ):
        mesh = triangle_with_uv_normal_color
        self.save_and_load(mesh)

    def test_single_triangle_with_attributes_explicit(
        self, triangle_with_uv_normal_color
    ):
        mesh = triangle_with_uv_normal_color
        attr_ids = mesh.get_matching_attribute_ids()
        assert len(attr_ids) != 0
        self.save_and_load(mesh, attr_ids)

    def test_mesh_string_conversion(self, triangle_with_uv_normal_color):
        mesh = triangle_with_uv_normal_color

        ply_data = lagrange.io.mesh_to_string(mesh, "ply")
        mesh2 = lagrange.io.string_to_mesh(ply_data)
        assert_same_vertices_and_facets(mesh, mesh2)

        obj_data = lagrange.io.mesh_to_string(mesh, "obj")
        mesh2 = lagrange.io.string_to_mesh(obj_data)
        assert_same_vertices_and_facets(mesh, mesh2)

        msh_data = lagrange.io.mesh_to_string(mesh, "msh")
        mesh2 = lagrange.io.string_to_mesh(msh_data)
        assert_same_vertices_and_facets(mesh, mesh2)

        gltf_data = lagrange.io.mesh_to_string(mesh, "gltf")
        mesh2 = lagrange.io.string_to_mesh(gltf_data)
        assert_same_vertices_and_facets(mesh, mesh2)

        gltf_data = lagrange.io.mesh_to_string(mesh, "glb")
        mesh2 = lagrange.io.string_to_mesh(gltf_data)
        assert_same_vertices_and_facets(mesh, mesh2)

    def test_io_with_selected_attributed(self, triangle):
        mesh = triangle
        selected_attributes = [
            mesh.create_attribute(
                "_vertex_id",
                element=lagrange.AttributeElement.Vertex,
                usage=lagrange.AttributeUsage.Scalar,
                initial_values=np.array([0, 1, 2], dtype=np.float32),
            ),
            mesh.create_attribute(
                "_facet_id",
                element=lagrange.AttributeElement.Facet,
                usage=lagrange.AttributeUsage.Scalar,
                initial_values=np.array([0], dtype=np.float32),
            ),
        ]

        for mesh_format in ["ply", "msh", "gltf", "glb"]:
            data = lagrange.io.mesh_to_string(
                mesh, mesh_format, selected_attributes=selected_attributes
            )
            mesh2 = lagrange.io.string_to_mesh(data)
            assert_same_vertices_and_facets(mesh, mesh2)
            assert mesh2.has_attribute("_vertex_id")
            assert mesh2.has_attribute("_facet_id")
